README.rst
This repository is archived and read only.

If you want to unarchive it, then post to the [Admin & Infrastructure (AI) Team category on the Plone Community Forum](https://community.plone.org/c/aiteam/55).

Introduction
============

plone.app.jquerytools adds `jquery.tools`_ and some related
overlay and form-handling JavaScript libraries to Plone.

``plone.app.jquerytools`` was developed for Plone 4. Earlier versions could
be used in Plone 3.x by adding a zcml slug and running its
GS Setup extension profile. This version requires Plone 4.x.

Plone developers will wish to use ``plone.app.jquerytools`` to provide DOM
popups, particularly if they require AJAX. There is extensive support
for AJAX form posting. It's also useful for adding dateinput and forminput
widgets that are HTML5 compatible.

The AJAX and AJAX-form support is Plone-specific, and is documented in this module.
Dateinput, rangeinput, accordions, and tooltips are all unchanged from
jquerytools, and the jquerytools docs should be consulted.

plone.app.jquerytools uses a jQuery 1.9+ compatible fork of jquerytools, located at: https://github.com/collective/jquerytools

.. contents::


Development
===========

Since version 2.0 we use a Grunt based build system. To update the package, you
have to setup Grunt like descibed below (this works on Linux). You need ``npm``
and ``git`` installed.

If grunt-cli isn't installed, do::

    $ sudo npm install -g grunt-cli

Then install the rest locally::

    $ npm install

For more information see: http://gruntjs.com/getting-started

To update the repository, simply run::

    $ grunt

.. note::
    You need git to be installed.


Available resources
===================

`jquery.tools`_ plugins and widgets are packed into Zope browser resources:

``++resource++plone.app.jquerytools.js``
    Default plugins and widgets used by plone. This resource is enabled
    by default with ``plone.app.jquerytools:default`` profile.

    Included scripts: `overlay.js`_, `scrollable.js`_, `tabs.js`_,
    `toolbox.history.js`_, `toolbox.expose.js`_, `tooltip.js`_

``++resource++plone.app.jquerytools.plugins.js``
    Additional plugin and widgets which does not take much space and for
    this reason are packed together. This plugins are not enabled by
    default.

    Included scripts: `overlay.apple.js`_, `scrollable.autoscroll.js`_,
    `scrollable.navigator.js`_, `tabs.slideshow.js`_, `toolbox.flashembed.js`_,
    `toolbox.mousewheel.js`_, `tooltip.dynamic.js`_, `tooltip.slide.js`_

``++resource++plone.app.jquerytools.dateinput.js`` ``++resource++plone.app.jquerytools.dateinput.css``
    `jquerytools dateinput`_ widget with style from `first demo`_. Both
    scripts are added to portal_javascript and portal_css but disabled by
    default.

``++resource++plone.app.jquerytools.rangeinput.js``
    `jquerytools rangeinput`_ widget. Added to portal_javascript tool, but
    disabled by default.

``++resource++plone.app.jquerytools.validator.js``
    `jquerytools validator`_ script, which should help you with nice
    validation of your forms. Added to portal_javascript tool, but
    disabled byt default.

``++resource++plone.app.jquerytools.form.js``
    Integrates the `jquery form plugin`_ to add support for AJAX form
     handling. More about this below.

``++resource++plone.app.jquerytools.overlayhelpers.js``
    not yet minimized) and ``++resource++plone.app.jquerytools.overlays.css``
    (Size: 1.9KB, not yet minimized)

    Adds helper code for loading overlays dynamically and for handling AJAX
    forms based on existing pages with minimal setup. More about this in
    instructions below.

JS resources are minified, but uncompressed versions are available in
plone/app/jquerytools/browser for reading/debugging purposes. To use them
for debugging, edit plone/app/jquerytools/configure.zcml to temporarily
specify files ending with .js rather than .min.js.


Overlay helpers
===============

plone.app.jquerytools provides a helper for handling various kinds of dynamic
overlays, including overlays with forms you wish handled by AJAX.

The helper, jQuery.fn.prepOverlay, is a jQuery-style function: it should be
used as a method of a jQuery selection object. The selection object is always
a selection of trigger elements.

prepOverlay should be passed one parameter: a options object, which will often
be constructed as a JavaScript literal object.


Examples
--------

Let's say, for example, that you want to make clicking on news-item photos
open a lightbox-style larger version of the image. To do this, you'll need to
specify:

* A jQuery style selector for a Plone element, e.g., ".newsImageContainer a"

* "image" for the load method ("ajax" and "iframe" are other alternatives)

* A regular expression search/replace to transform the href or src URL.
  In this example, we're changing the URL to point to the preview-sized
  image. So, our search/replace pair is "/image_view_fullscreen"
  and "_preview".

* You could also specify additional overlay configuration parameters.

The code::

    $('.newsImageContainer a')
        .prepOverlay({
             subtype: 'image',
             urlmatch: '/image_view_fullscreen$',
             urlreplace: '_preview'
            });

What if you want the lightbox-style larger image to open upon doubleclick or
hover, instead? Just add a "event" attribute to override the default::

    $('.newsImageContainer a')
        .prepOverlay({
             subtype: 'image',
             urlmatch: '/image_view_fullscreen$',
             urlreplace: '_preview',
             event: 'dblclick'
            });

The optional "event" attribute takes one of three values: either the default
"click", "dblclick", or "hover".

Another quick example, one that provides full-image popups for images placed
via kupu::

    $('img.image-right, img.image-left, img.image-inline')
        .prepOverlay({
            subtype: 'image',
            urlmatch: '/image_.+$',
            urlreplace: ''
            });

What's different? We're targeting <img ... /> tags, which don't have href
attributes. The helper automatically picks up the target URL from the src
attribute, so that we can have a popup view of image elements that aren't
linked to that view. Note also that we're using a real regular expression
in the search/replace so that we can strip off image_preview, image_mini, etc.

And, a configuration to put the site map in an iframe popup with expose
settings, picking up the target from an href::

    $('#siteaction-sitemap a')
        .prepOverlay({
            subtype: 'iframe',
            config: {expose:{color:'#00f'}}
            });

Options
-------

The complete options list:

subtype
  'image' | 'iframe' | 'ajax'
urlmatch:
  Regular expression for a portion of the target URL. Target
  URL is determined by checking href, src or action attributes.
urlreplace
  Replacement expression for the matched expression.
filter (ajax only)
  the jQuery selector used to find the elements of
  the ajax loaded resource that you wish to use in the overlay.
width
  Width of the popup. Defaults to 60%. Overriden by image width
  for image overlays. Percentages are computed against window width,
  not parent.
cssclass
  A custom css class to apply to the overlay. Ignored
  for inline overlays.
config
  jQuery Tools configuration options in a dictionary.

For AJAX overlay forms, add the following, form-oriented, options:

formselector
  Used to specify the JQuery selector for any
  forms inside the loaded content that you want to be handled
  inside the overlay by doing an AJAX load to get the overlay
  content.

  When a form is submitted, the overlay handler checks the response
  for formselector. If it's found, the result is displayed in the
  overlay and form handlers are bound. If not, the 'noform' action
  is carried out.

noform
  the action to take if an ajax form is submitted and the returned
  content has nothing matching the formselector. Available actions include
  'close' to simply close the overlay, 'reload' to reload the page, and
  'redirect' to redirect to another page. If you choose 'redirect', you
  must specify the URL in the redirect option. Default
  action is to display the filtered response in the popup.

  You may also supply as the 'noform' argument a
  callback function that returns one of these strings. The overlay helper
  will call the function with the overlay element as an argument.

closeselector
  use this to specify a JQuery selector that will be used
  to find elements within the overlay that should close the overlay if
  clicked. The most obvious example is a form's cancel button.

redirect
  if you specify 'redirect' for the noform action, use the
  redirect option to specify the full target URL. You may also supply a
  callback function that returns a URL. The overlay helper will call the
  function with the overlay element and the response text as arguments.

beforepost
  you may specify a function that will be called before the
  AJAX form posting. This callback will be passed the jQuery-wrapped form
  and the serialized form data. Return true if you wish the AJAX form
  handler to handle the event; return false if you wish to cancel the
  submit.

afterpost
  you may specify a function that will be called immediately
  after the AJAX load of the post response. The function will be passed an

CHANGES.rst
Changelog
=========

1.9.6 (unreleased)
------------------

Breaking changes:

- *add item here*

New features:

- Add uninstall profile.
  [thet]

Bug fixes:

- *add item here*


1.9.5 (2019-01-08)
------------------

Bug fixes:

- More quote fixes for scss
  [agitator]


1.9.4 (2018-01-17)
------------------

Bug fixes:

- Please pyScss by adding quotes in the urls. Makes it compatible with Plone 5.1.
  [thomasdesvenain]


1.9.3 (2017-10-11)
------------------

Bug fixes:

- Disabled overlays.css to match upgradestep.
  [jaroel]


1.9.2 (2017-06-28)
------------------

Bug fixes:

- Update ``Gruntfile.js`` with master branch of collective/jquerytools.
  Update docs with info about our jQuery Tools fork.
  [thet]


1.9.1 (2017-03-23)
------------------

Bug fixes:

- Updated jquery.form and jquery.tools to latest versions.
  [msom]

1.9.0 (2016-12-02)
------------------

New features:

- An overlay registered by the prepOverlay function can now be optionally be
  triggered by a hover or doubleclick event, instead of click.
  [petri]


1.8.0 (2016-08-12)
------------------

New features:

- The prepOverlay function can be used on a form instead of a link. When a
  button of this form is clicked, the form action attribute is used to retrieve
  the overlay. (This was already the case in previous versions, What's new is
  the following.) If method="POST", the overlay will be fetched via a POST with
  the form payload instead of GET.
  The use of a POST instead of GET may be useful to avoid the url length limit.
  [vincentfretin]


1.7.1 (2016-06-27)
------------------

- Proper clean setup of jquery.form
  [gotcha]


1.7.0 (2015-04-29)
------------------

- Remove the development profile. Unminified resources are still provided. If
  you need them, include them in your own profile.
  [thet]

- Move jQueryTools tooltip into separate file, so that it can be selectively
  disabled.
  [thet]


1.6.2 (2014-11-04)
------------------

- Revert "check each tooltip for existence, fixes issue 349". Fixes broken
  Tooltip.
  [vincentfretin, thet]


1.6.1 (2014-09-11)
------------------

- Fix dateinput (change event was not triggered anymore).
  [thomasdesvenain]


1.6.0 (2014-07-10)
------------------

- Ensure jQuery 1.7+, including 1.9 compatibility by using the dev branch of
  jquerytools and applying further fixes. We now use a fork of jquerytools at:
  https://github.com/collective/jquerytools/tree/dev
  [thet]

- Fix ``overlayhelpers.js`` to work with jQuery 1.7 and 1.9.
  jQuery.buildFragment has a slightly changed API. While it accepts a list of
  context objects in jQuery 1.7, it expects an explicit context object in
  jQuery 1.9. See::

    - https://github.com/jquery/jquery/blob/1.7/src/manipulation.js#L465
    - https://github.com/jquery/jquery/blob/1.7/src/manipulation.js#L472
    - https://github.com/jquery/jquery/blob/1.9.1/src/manipulation.js#L617

  [thet]

- Switched to Grunt based build system.
  [thet]

- Switch test infrastructure to plone.app.testing.
  [sdelcourt]


1.5.7 (2014-02-23)
------------------

- Fix a test isolation issue.
  [davisagli]

- Install plone.app.jquerytools.js after plone_javasscript_variables.js
  in new instances.
  [bloodbare]


1.5.6 (2013-08-13)
------------------

- In the date picker, fire the change event on the input
  so it doesn't bubble up and cause an error in jquery 1.8.
  [davisagli]

- Backport pull-req https://github.com/jquerytools/jquerytools/pull/873 and
  apply in other jquerytools files too: Fix API of outerWidth and outerHeight
  to ensure jQuery 1.8.x compatibility. This change is backwards compatible at
  least to 1.4, since jquerytools used the API incorrectly. See jQuery 1.4 API
  doc:
  http://web.archive.org/web/20100210083046/http://api.jquery.com/outerWidth/
  [thet]

1.5.5 (2013-03-18)
------------------

- Set z-index for dateinput calendar popup so that it will display in
  an overlay context.
  [smcmahon]

- Pass the pbo object as third parameter to the redirect callback. This
  provides the same flexibility for redirect that was added for noform in
  1.5.4.
  [izak]


1.5.4 (2013-03-04)
------------------

- Now execute inline scripts in ajax overlay.
  [vincentfretin]

- Give the pbo object as second parameter for noform callback. You can access
  everything from it, for example the overlay trigger pbo.source.
  [vincentfretin]

- Give the disconnected "el" jQuery object (the div created with the html
  response) instead of "this" (the request object) to the noform and redirect
  callbacks. This fixes the noformerrorshow callback from popupforms.js in the
  Products.CMFPlone package.
  [vincentfretin]


plone/app/jquerytools/browser/overlayhelpers.js
/*****************

   jQuery Tools overlay helpers.

   Copyright © 2010, The Plone Foundation
   Licensed under the GPL, see LICENSE.txt for details.

*****************/

/*jslint browser: true */
/*global jQuery, ajax_noresponse_message, close_box_message, window */

// Name space object for pipbox
var pb = {spinner: {}, overlay_counter: 1};

jQuery.tools.overlay.conf.oneInstance = false;

(function($) {
    // override jqt default effect to take account of position
    // of parent elements
    jQuery.tools.overlay.addEffect('default',
        function(pos, onLoad) {
            var conf = this.getConf(),
                 w = $(window),
                 ovl = this.getOverlay(),
                 op = ovl.parent().offsetParent().offset();

            if (!conf.fixed)  {
                pos.top += w.scrollTop() - op.top;
                pos.left += w.scrollLeft() - op.left;
            }

            pos.position = conf.fixed ? 'fixed' : 'absolute';
            ovl.css(pos).fadeIn(conf.speed, onLoad);

        }, function(onClose) {
            this.getOverlay().fadeOut(this.getConf().closeSpeed, onClose);
        }
    );

    pb.spinner.show = function () {
        $('body').css('cursor', 'wait');
    };
    pb.spinner.hide = function () {
        $('body').css('cursor', '');
    };
}(jQuery));


jQuery(function ($) {

    /******
        $.fn.prepOverlay jQuery plugin to inject overlay target into DOM and
        annotate it with the data we'll need in order to display it.
    ******/
    $.fn.prepOverlay = function (pba) {
        return this.each(function () {
            var o, pbo, config, onBeforeLoad, onLoad, src, parts;

            o = $(this);

            // copy options so that it's not just a reference
            // to the parameter.
            pbo = $.extend(true, {'width':'60%'}, pba);

            // set overlay configuration
            config = pbo.config || {};

            // set onBeforeLoad handler
            onBeforeLoad = pb[pbo.subtype];
            if (onBeforeLoad) {
                config.onBeforeLoad = onBeforeLoad;
            }
            onLoad = config.onLoad;
            config.onLoad = function () {
                if (onLoad) {
                    onLoad.apply(this, arguments);
                }
                pb.fi_focus(this.getOverlay());
            };

            // be promiscuous, pick up the url from
            // href, src or action attributes
            src = o.attr('href') || o.attr('src') || o.attr('action');

            // translate url with config specifications
            if (pbo.urlmatch) {
                src = src.replace(new RegExp(pbo.urlmatch), pbo.urlreplace);
            }

            if (pbo.subtype === 'inline') {
                // we're going to let tools' overlay do all the real
                // work. Just get the markers in place.
                src = src.replace(/^.+#/, '#');
                $("[id='" + src.replace('#', '') + "']").addClass('overlay');
                o.removeAttr('href').attr('rel', src);
                // use overlay on the source (clickable) element
                o.overlay();
            } else {
                // save various bits of information from the pbo options,
                // and enable the overlay.

                // this is not inline, so in one fashion or another
                // we'll be loading it via the beforeLoad callback.
                // create a unique id for a target element
                pbo.nt = 'pb_' + pb.overlay_counter;
                pb.overlay_counter += 1;

                pbo.selector = pbo.filter || pbo.selector;
                if (!pbo.selector) {
                    // see if one's been supplied in the src
                    parts = src.split(' ');
                    src = parts.shift();
                    pbo.selector = parts.join(' ');
                }

                pbo.src = src;
                pbo.config = config;
                pbo.source = o;

                // remove any existing overlay and overlay handler
                pb.remove_overlay(o);

                // save options on trigger element
                o.data('pbo', pbo);

                // mark the source with a rel attribute so we can find
                // the overlay, and a special class for styling
                o.attr('rel', '#' + pbo.nt);
                o.addClass('link-overlay');

                // for some subtypes, we're setting click handlers
                // and attaching overlay to the target element. That's
                // so we'll know the dimensions early.
                // Others, like iframe, just use overlay.

                register_handler = function(evt, o, hnd) {
                   switch (evt) {
                      case undefined:
                         o.click(hnd);
                         break;
                      case 'click':
                         o.click(hnd);
                         break;
                      case 'hover':
                         o.hover(hnd);
                         break;
                      case 'dblclick':
                         o.dblclick(hnd);
                         break;
                      default:
                         throw "Unsupported event";
                   }
                }

                switch (pbo.subtype) {
                case 'image':
                    register_handler(pbo.event, o, pb.image_click);
                    break;
                case 'ajax':
                    register_handler(pbo.event, o, pb.ajax_click);
                    break;
                case 'iframe':
                    pb.create_content_div(pbo);
                    o.overlay(config);
                    break;
                default:
                    throw "Unsupported overlay type";
                }

                // in case the click source wasn't
                // already a link.
                o.css('cursor', 'pointer');
            }
        });
    };


    /******
        pb.remove_overlay
        Remove the overlay and handler associated with a jquery wrapped
        trigger object
    ******/
    pb.remove_overlay = function (o) {
        var old_data = o.data('pbo');
        if (old_data) {
            switch (old_data.subtype) {
                case 'image':
                    o.unbind('click', pb.image_click);
                    break;
                case 'ajax':
                    o.unbind('click', pb.ajax_click);
                    break;
                default:
                    // it's probably the jqt overlay click handler,
                    // but we don't know the handler and are forced
                    // to do a generic unbind of click handlers.
                    o.unbind('click');
                    break;
            }
            if (old_data.nt) {
                $('#' + old_data.nt).remove();
            }
        }
    };


    /******
        pb.create_content_div
        create a div to act as an overlay; append it to
        the body; return it
    ******/
    pb.create_content_div = function (pbo, trigger) {
        var content,
            close_message,
            pbw = pbo.width;

        if (typeof close_box_message === 'undefined') {
            close_message = 'Close this box.';
        } else {
            close_message = close_box_message;
        }

        content = $(
            '<div id="' + pbo.nt +
            '" class="overlay overlay-' + pbo.subtype +
            ' ' + (pbo.cssclass || '') +
            '"><div class="close"><a href="#" class="hiddenStructure" title="Close this box">' +
            close_message +
            '</a></div></div>');

        content.data('pbo', pbo);

        // if a width option is specified, set it on the overlay div,
        // computing against the window width if a % was specified.
        if (pbw) {
            if (pbw.indexOf('%') > 0) {
                content.width(parseInt(pbw, 10) / 100 * $(window).width());
            } else {
                content.width(pbw);
            }
        }

        // add the target element at the end of the body.
        content.appendTo($("body"));

        return content;
    };


    /******
        pb.image_click
        click handler for image loads
    ******/
    pb.image_click = function (event) {
        var ethis, content, api, img, el, pbo;

        ethis = $(this);
        pbo = ethis.data('pbo');

        // find target container
        content = $(ethis.attr('rel'));
        if (!content.length) {
            content = pb.create_content_div(pbo);
            content.overlay(pbo.config);
        }
        api = content.overlay();

        // is the image loaded yet?
        if (content.find('img').length === 0) {
            // load the image.
            if (pbo.src) {
                pb.spinner.show();

                // create the image and stuff it
                // into our target
                img = new Image();
                img.src = pbo.src;
                el = $(img);
                content.append(el.addClass('pb-image'));

                // Now, we'll cause the overlay to
                // load when the image is loaded.
                el.load(function () {
                    pb.spinner.hide();
                    api.load();
                });

            }
        } else {
            api.load();
        }

        return false;
    };


    /******
        pb.fi_focus
        First-input focus inside $ selection.
    ******/
    pb.fi_focus = function (jqo) {
        if (! jqo.find("form div.error :input:first").focus().length) {
            jqo.find("form :input:visible:first").focus();
        }
    };


    /******
        pb.ajax_error_recover
        jQuery's ajax load function does not load error responses.
        This routine returns the cooked error response.
    ******/
    pb.ajax_error_recover = function (responseText, selector) {
        var tcontent = $('<div/>').append(
                responseText.replace(/<script(.|\s)*?\/script>/gi, ""));
        return selector ? tcontent.find(selector) : tcontent;
    };


    /******
        pb.add_ajax_load
        Adds a hidden ajax_load input to form
    ******/
    pb.add_ajax_load = function (form) {
        if (form.find('input[name=ajax_load]').length === 0) {
            form.prepend($('<input type="hidden" name="ajax_load" value="' +
                (new Date().getTime()) +
                '" />'));
        }
    };

    /******
        pb.prep_ajax_form
        Set up form with ajaxForm, including success and error handlers.
    ******/
    pb.prep_ajax_form = function (form) {
        var ajax_parent = form.closest('.pb-ajax'),
            data_parent = ajax_parent.closest('.overlay-ajax'),
            pbo = data_parent.data('pbo'),
            formtarget = pbo.formselector,
            closeselector = pbo.closeselector,
            beforepost = pbo.beforepost,
            afterpost = pbo.afterpost,
            noform = pbo.noform,
            api = data_parent.overlay(),
            selector = pbo.selector,
            options = {};

        options.beforeSerialize = function () {
            pb.spinner.show();
        };

        if (beforepost) {
            options.beforeSubmit = function (arr, form, options) {
                return beforepost(form, arr, options);
            };
        }
        options.success = function (responseText, statusText, xhr, form) {
            $(document).trigger('formOverlayStart', [this, responseText, statusText, xhr, form]);
            // success comes in many forms, some of which are errors;
            //

            var el, myform, success, target, scripts = [], filteredResponseText;

            success = statusText === "success" || statusText === "notmodified";

            if (! success) {
                // The responseText parameter is actually xhr
                responseText = responseText.responseText;
            }
            // strip inline script tags
            filteredResponseText = responseText.replace(/<script(.|\s)*?\/script>/gi, "");

            // create a div containing the optionally filtered response
            el = $('<div />').append(
                selector ?
                    // a lesson learned from the jQuery source: $(responseText)
                    // will not work well unless responseText is well-formed;
                    // appending to a div is more robust, and automagically
                    // removes the html/head/body outer tagging.
                    $('<div />').append(filteredResponseText).find(selector)
                    :
                    filteredResponseText
                );

            // afterpost callback
            if (success && afterpost) {
                afterpost(el, data_parent);
            }

            myform = el.find(formtarget);
            if (success && myform.length) {
                ajax_parent.empty().append(el);

                // execute inline scripts
                try {
                    // jQuery 1.7
                    $.buildFragment([responseText], [document], scripts);
                } catch (e) {
                    // jQuery 1.9
                    $.buildFragment([responseText], document, scripts);
                }
                if (scripts.length) {
                    $.each(scripts, function() {
                        $.globalEval( this.text || this.textContent || this.innerHTML || "" );
                    });
                }

                // This may be a complex form.
                if ($.fn.ploneTabInit) {
                    el.ploneTabInit();
                }
                pb.fi_focus(ajax_parent);

                pb.add_ajax_load(myform);
                // attach submit handler with the same options
                myform.ajaxForm(options);

                // attach close to element id'd by closeselector
                if (closeselector) {
                    el.find(closeselector).click(function (event) {
                        api.close();
                        return false;
                    });
                }
                $(document).trigger('formOverlayLoadSuccess', [this, myform, api, pb, ajax_parent]);
            } else {
                // there's no form in our new content or there's been an error
                if (success) {
                    if (typeof noform === "function") {
                        // get action from callback
                        noform = noform(el, pbo);
                    }
                } else {
                    noform = statusText;
                }


                switch (noform) {
                case 'close':
                    api.close();
                    break;
                case 'reload':
                    api.close();
                    pb.spinner.show();
                    // location.reload results in a repost
                    // dialog in some browsers; very unlikely to
                    // be what we want.
                    location.replace(location.href);
                    break;
                case 'redirect':
                    api.close();
                    pb.spinner.show();
                    target = pbo.redirect;
                    if (typeof target === "function") {
                        // get target from callback
                        target = target(el, responseText, pbo);
                    }
                    location.replace(target);
                    break;
                default:
                    if (el.children()) {
                        // show what we've got
                        ajax_parent.empty().append(el);
                    } else {
                        api.close();
                    }
                    break;
                }
                $(document).trigger('formOverlayLoadFailure', [this, myform, api, pb, ajax_parent, noform]);
            }
            pb.spinner.hide();
        };
        // error and success callbacks are the same
        options.error = options.success;

        pb.add_ajax_load(form);
        form.ajaxForm(options);
    };


    /******
        pb.ajax_click
        Click handler for ajax sources. The job of this routine
        is to do the ajax load of the overlay element, then
        call the JQT overlay loader.
    ******/
    pb.ajax_click = function (event) {
        var ethis = $(this),
            pbo,
            content,
            api,
            src,
            el,
            selector,
            formtarget,
            closeselector,
            sep,
            scripts = [],
            e;

        e = $.Event();
        e.type = "beforeAjaxClickHandled";
        $(document).trigger(e, [this, event]);
        if (e.isDefaultPrevented()) { return false; }

        pbo = ethis.data('pbo');

        content = pb.create_content_div(pbo, ethis);
        // pbo.config.top = $(window).height() * 0.1 - ethis.offsetParent().offset().top;
        content.overlay(pbo.config, ethis);
        api = content.overlay();
        src = pbo.src;
        selector = pbo.selector;
        formtarget = pbo.formselector;
        closeselector = pbo.closeselector;

        pb.spinner.show();

        // prevent double click warning for this form
        $(this).find("input.submitting").removeClass('submitting');

        el = $('div.pb-ajax', content);
        if (el.length === 0) {
            el = $('<div class="pb-ajax" />');
            content.append(el);
        }
        if (api.getConf().fixed) {
            // don't let it be over 75% of the viewport's height
            el.css('max-height', Math.floor($(window).height() * 0.75));
        }

        // affix a random query argument to prevent
        // loading from browser cache
        sep = (src.indexOf('?') >= 0) ? '&': '?';
        src += sep + "ajax_load=" + (new Date().getTime());

        // add selector, if any
        if (selector) {
            src += ' ' + selector;
        }

        // set up callback to be used whenever new contents are loaded
        // into the overlay, to prepare links and forms to stay within
        // the overlay
        el[0].handle_load_inside_overlay = function(responseText, errorText) {
            var ele, target;
            ele = $(this);

            if (errorText === 'error') {
                ele.append(pb.ajax_error_recover(responseText, selector));
            } else if (!responseText.length) {
                ele.append(ajax_noresponse_message || 'No response from server.');
            }

            // a non-semantic div here will make sure we can
            // do enough formatting.
            ele.wrapInner('<div />');

            // add the submit handler if we've a formtarget
            if (formtarget) {
                target = ele.find(formtarget);
                if (target.length > 0) {
                    pb.prep_ajax_form(target);
                }
            }

            // if a closeselector has been specified, tie it to the overlay's
            // close method via closure
            if (closeselector) {
                ele.find(closeselector).click(function (event) {
                    api.close();
                    return false;
                });
            }

            // execute inline scripts
            try {
                // jQuery 1.7
                $.buildFragment([responseText], [document], scripts);
            } catch (e) {
                // jQuery 1.9
                $.buildFragment([responseText], document, scripts);
            }
            if (scripts.length) {
                $.each(scripts, function() {
                    $.globalEval( this.text || this.textContent || this.innerHTML || "" );
                });
            }

            // This may be a complex form.
            if ($.fn.ploneTabInit) {
                ele.ploneTabInit();
            }

            // remove element on close so that it doesn't congest the DOM
            api.onClose = function () {
                content.remove();
            };
            $(document).trigger('loadInsideOverlay', [this, responseText, errorText, api]);
        };

        // and load the div
        var data = null;
        var form = $(pbo.source.context);
        var method = form.attr('method');
        if (method && method.toLowerCase() == 'post') {
            data = form.serializeArray();
        }
        // load will do a POST instead of GET if data is not null
        el.load(src, data, function (responseText, errorText) {
            // post-process the overlay contents
            el[0].handle_load_inside_overlay.apply(this, [responseText, errorText]);

            // Now, it's all ready to display; hide the
            // spinner and call JQT overlay load.
            pb.spinner.hide();
            api.load();

            return true;
        });

        // don't do the default action
        return false;
    };


    /******
        pb.iframe
        onBeforeLoad handler for iframe overlays.

        Note that the spinner is handled a little differently
        so that we can keep it displayed while the iframe's
        content is loading.
    ******/
    pb.iframe = function () {
        var content, pbo;

        pb.spinner.show();

        content = this.getOverlay();
        pbo = this.getTrigger().data('pbo');

        if (content.find('iframe').length === 0 && pbo.src) {
            content.append(
                '<iframe src="' + pbo.src + '" width="' +
                 content.width() + '" height="' + content.height() +
                 '" onload="pb.spinner.hide()"/>');
        } else {
            pb.spinner.hide();
        }
        return true;
    };

    // $('.newsImageContainer a')
    //     .prepOverlay({
    //          subtype: 'image',
    //          urlmatch: '/image_view_fullscreen$',
    //          urlreplace: '_preview'
    //         });

});
